In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.linear_model import LogisticRegression

In [2]:
from google.colab import files
uploaded = files.upload()

Saving jobclassinfo2.csv to jobclassinfo2.csv


In [3]:
train = pd.read_csv("jobclassinfo2.csv")

In [4]:
train.head()

ID  JobFamily    JobFamilyDescription  JobClass JobClassDescription  \
0   1          1  Accounting And Finance         1        Accountant I   
1   2          1  Accounting And Finance         2       Accountant II   
2   3          1  Accounting And Finance         3      Accountant III   
3   4          1  Accounting And Finance         4       Accountant IV   
4   5          2  Administrative Support         5     Admin Support I   

   PayGrade  EducationLevel  Experience  OrgImpact  ProblemSolving  \
0         5               3           1          3               3   
1         6               4           1          5               4   
2         8               4           2          6               5   
3        10               5           5          6               6   
4         1               1           0          1               1   

   Supervision  ContactLevel  FinancialBudget    PG  
0            4             3                5  PG05  
1            5             7                7  PG06  
2            6             7               10  PG08  
3            7             8               11  PG10  
4            1             1                1  PG01

In [5]:
print(train.shape)

(66, 14)


In [6]:
print(train.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ID                    66 non-null     int64 
 1   JobFamily             66 non-null     int64 
 2   JobFamilyDescription  66 non-null     object
 3   JobClass              66 non-null     int64 
 4   JobClassDescription   66 non-null     object
 5   PayGrade              66 non-null     int64 
 6   EducationLevel        66 non-null     int64 
 7   Experience            66 non-null     int64 
 8   OrgImpact             66 non-null     int64 
 9   ProblemSolving        66 non-null     int64 
 10  Supervision           66 non-null     int64 
 11  ContactLevel          66 non-null     int64 
 12  FinancialBudget       66 non-null     int64 
 13  PG                    66 non-null     object
dtypes: int64(11), object(3)
memory usage: 7.3+ KB
None


In [7]:
print(train.isna().sum())

ID                      0
JobFamily               0
JobFamilyDescription    0
JobClass                0
JobClassDescription     0
PayGrade                0
EducationLevel          0
Experience              0
OrgImpact               0
ProblemSolving          0
Supervision             0
ContactLevel            0
FinancialBudget         0
PG                      0
dtype: int64


In [8]:
from sklearn import preprocessing
label_encoder = preprocessing.LabelEncoder() 
label_encoder.fit(train['JobFamilyDescription']) 
train['JobFamilyDescription'] = label_encoder.transform(train['JobFamilyDescription'])
label_encoder.fit(train['JobClassDescription']) 
train['JobClassDescription'] = label_encoder.transform(train['JobClassDescription'])
label_encoder.fit(train['PG']) 
train['PG'] = label_encoder.transform(train['PG'])
train.head()

ID  JobFamily  JobFamilyDescription  JobClass  JobClassDescription  \
0   1          1                     0         1                    0   
1   2          1                     0         2                    1   
2   3          1                     0         3                    2   
3   4          1                     0         4                    3   
4   5          2                     1         5                    4   

   PayGrade  EducationLevel  Experience  OrgImpact  ProblemSolving  \
0         5               3           1          3               3   
1         6               4           1          5               4   
2         8               4           2          6               5   
3        10               5           5          6               6   
4         1               1           0          1               1   

   Supervision  ContactLevel  FinancialBudget  PG  
0            4             3                5   4  
1            5             7                7   5  
2            6             7               10   7  
3            7             8               11   9  
4            1             1                1   0

In [9]:
features = pd.Series(['ID','JobFamilyDescription', 'JobClass', 'JobClassDescription', 'PayGrade', 'EducationLevel', 'Experience', 'OrgImpact', 'ProblemSolving', 'Supervision', 'ContactLevel', 'FinancialBudget', 'PG'])
from sklearn.linear_model import LinearRegression
X = train[['ID', 'JobFamily', 'JobFamilyDescription', 'JobClass',
       'JobClassDescription', 'PayGrade', 'EducationLevel', 'Experience',
       'OrgImpact', 'ProblemSolving', 'Supervision', 'ContactLevel',
       'FinancialBudget', 'PG']].values
y = train['JobFamily'].values
slr = LinearRegression() 
slr.fit(X, y)
coeff_df = pd.DataFrame(slr.coef_, columns=['Coefficient'])  
coeff_df['features'] = features

coeff_df

Coefficient              features
0   2.964385e-16                    ID
1   5.000000e-01  JobFamilyDescription
2   5.000000e-01              JobClass
3  -4.996004e-16   JobClassDescription
4  -5.551115e-17              PayGrade
5   3.122502e-17        EducationLevel
6   6.938894e-18            Experience
7  -1.734723e-17             OrgImpact
8   1.526557e-16        ProblemSolving
9  -1.604619e-16           Supervision
10 -9.020562e-17          ContactLevel
11 -1.960238e-16       FinancialBudget
12 -3.126839e-16                    PG
13 -4.128642e-16                   NaN

In [10]:
slr.intercept_

0.5000000000000098

0.5000000000000098 + (2.964385e-16 ID) + (5.000000e-01 JobFamilyDescription) + (5.000000e-01 JobClass) + (-4.996004e-16 JobClassDescription) + (-5.551115e-17 PayGrade) + (3.122502e-17 EducationLevel) + (6.938894e-18 Experience) + (-1.734723e-17 OrgImpact) + (1.526557e-16 ProblemSolving) + (-1.604619e-16 Supervision) + (-9.020562e-17	ContactLevel) + (-1.960238e-16 FinancialBudget) + (-3.126839e-16 PG) + (-4.128642e-16 NaN)

In [41]:
from sklearn.model_selection import train_test_split

# train - на тренировочные и тестовые
X = train.drop(['JobFamily'], axis=1)
y = train['JobFamily']

X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.15)

model = LogisticRegression()
model.fit(X_train, y_train)
answers_pred = model.predict(X_test)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [42]:
from sklearn.metrics import accuracy_score
# Как оценить насколько хорошо работает ваша модель?
# answers_pred - ответы которые вернула модель для X_test
# y_test - это правильные ответы для X_test
print(accuracy_score(y_test,answers_pred))

0.8


In [43]:
print(X_train.shape)
print(y_train.shape)

(56, 13)
(56,)


In [44]:
predictions = model.predict(X)
predictions

array([ 1,  1,  1,  1,  2,  2,  2,  2,  2,  3,  3,  3,  4,  4,  4,  4,  4,
        4,  4,  4,  5,  5,  5, 10,  6,  7,  7,  7,  7,  7,  7,  7,  7,  7,
        7,  7,  5,  7,  8,  8,  8,  9,  9,  9, 10, 10, 10, 11, 11, 11, 11,
       12, 12, 12, 12, 12, 13, 15, 14, 14, 14, 15, 15, 15, 12, 15])